deployment/tls/tls.tf

```diff
@@ -0,0 +1,146 @@
+variable "resource_group_name" {
+  description = "Resource group name"
+  type        = "string"
+}
+
+variable "resource_group_location" {
+  description = "Resource group location"
+  type        = "string"
+}
+
+resource "random_string" "server" {
+  length  = 8
+  upper   = false
+  special = false
+  number  = false
+}
+
+locals {
+  fqdn   = "ion${random_string.server.result}.${var.resource_group_location}.cloudapp.azure.com"
+  prefix = "ion${random_string.server.result}"
+}
+
+resource "tls_private_key" "ca" {
+  algorithm = "RSA"
+  rsa_bits  = "2048"
+}
+
+resource "tls_self_signed_cert" "ca" {
+  key_algorithm   = "${tls_private_key.ca.algorithm}"
+  private_key_pem = "${tls_private_key.ca.private_key_pem}"
+
+  subject {
+    common_name  = "Ion CA"
+    organization = "Ion, Ltd"
+    country      = "GB"
+  }
+
+  validity_period_hours = 43800
+  is_ca_certificate     = true
+
+  allowed_uses = [
+    "key_encipherment",
+    "digital_signature",
+    "server_auth",
+    "client_auth",
+    "cert_signing",
+  ]
+}
+
+resource "tls_private_key" "server" {
+  algorithm = "RSA"
+  rsa_bits  = "2048"
+}
+
+resource "tls_cert_request" "server" {
+  key_algorithm   = "${tls_private_key.server.algorithm}"
+  private_key_pem = "${tls_private_key.server.private_key_pem}"
+
+  subject {
+    common_name  = "${local.fqdn}"
+    organization = "Ion, Ltd"
+    country      = "GB"
+  }
+
+  dns_names = ["${local.fqdn}"]
+}
+
+resource "tls_locally_signed_cert" "server" {
```

Comment:
This module is awesome, worthy of a blog post as probably really useful to others too